audio/audio_hw.c

```diff
@@ -396,9 +398,84 @@ static int out_set_parameters(struct audio_stream *stream, const char *kvpairs)
     return 0;
 }
 
-static char * out_get_parameters(const struct audio_stream *stream, const char *keys)
+static void str_parms_add_audio_format(struct str_parms* str_parms, 
+    const char* audio_parms, audio_format_t audio_format) {
+    char* format;
+
+    switch (audio_format) {
+    case AUDIO_FORMAT_PCM:
+        format = "AUDIO_FORMAT_PCM";
+        break;
+    case AUDIO_FORMAT_MP3:
+        format = "AUDIO_FORMAT_MP3";
+        break;
+    case AUDIO_FORMAT_AMR_NB:
+        format = "AUDIO_FORMAT_AMR_NB";
+        break;
+    case AUDIO_FORMAT_AMR_WB:
+        format = "AUDIO_FORMAT_AMR_WB";
+        break;
+    case AUDIO_FORMAT_AAC:
+        format = "AUDIO_FORMAT_AAC";
+        break;
+    case AUDIO_FORMAT_HE_AAC_V1:
+        format = "AUDIO_FORMAT_HE_AAC_V1";
+        break;
+    case AUDIO_FORMAT_HE_AAC_V2:
+        format = "AUDIO_FORMAT_HE_AAC_V2";
+        break;
+    case AUDIO_FORMAT_VORBIS:
+        format = "AUDIO_FORMAT_VORBIS";
+        break;
+    case AUDIO_FORMAT_MAIN_MASK:
+        format = "AUDIO_FORMAT_MAIN_MASK";
+        break;
+    case AUDIO_FORMAT_SUB_MASK:
+        format = "AUDIO_FORMAT_SUB_MASK";
+        break;
+    case AUDIO_FORMAT_PCM_16_BIT:
+        format = "AUDIO_FORMAT_PCM_16_BIT";
+        break;
+    case AUDIO_FORMAT_PCM_8_BIT:
+        format = "AUDIO_FORMAT_PCM_8_BIT";
+        break;
+    case AUDIO_FORMAT_PCM_32_BIT:
+        format = "AUDIO_FORMAT_PCM_32_BIT";
+        break;
+    case AUDIO_FORMAT_PCM_8_24_BIT:
+        format = "AUDIO_FORMAT_PCM_8_24_BIT";
+        break;
+    default:
+        format = "AUDIO_FORMAT_INVALID";
+    }
+
+    str_parms_add_str(str_parms, audio_parms, format);
+}
+
+static char* out_get_parameters(const struct audio_stream *stream, const char *keys)
 {
-    return strdup("");
+    ALOGV("+out_get_parameters(%p) '%s'", stream, keys);
+    
+    struct stream_out_common *out = (struct stream_out_common *)stream;
+    struct str_parms *query = str_parms_create_str(keys);
+    struct str_parms *reply = str_parms_create();
+    char value[BUFFER_SIZE];
+    char *str;
+
+    if (str_parms_get_str(query, AUDIO_PARAMETER_STREAM_SUP_FORMATS,
+        value, sizeof(value)) >= 0) {
+        str_parms_add_audio_format(reply, 
+            AUDIO_PARAMETER_STREAM_SUP_FORMATS, out->format);
+        str = strdup(str_parms_to_str(reply));
+    } else {
+        str = strdup("");
+    }
```

Comment:
This strdup("") isn't necessary. Can call str_parms_to_str(reply) outside the if. If nothing was added to reply it will return the appropriate empty string.

Reply:
Done.

Reply:
Done.